File: server/tools.py
"""
Adobe Premiere Pro MCP Server tools for project, sequence, and media management
"""

import logging
import functools
from typing import Dict, Any, Callable, Optional, TypeVar, Awaitable, Tuple

from websocket_server import get_server

logger = logging.getLogger(__name__)

# Type hint definitions
T = TypeVar('T')
ActionFunction = Callable[..., Awaitable[Tuple[str, Dict[str, Any]]]]

# Action processing decorator
def premiere_action(func: ActionFunction) -> Callable[..., Awaitable[Dict[str, Any]]]:
    """
    Decorator: Handles common processing for Premiere Pro actions
    - Error handling
    - Communication with WebSocket server
    - Response standardization
    """
    @functools.wraps(func)
    async def wrapper(*args, **kwargs) -> Dict[str, Any]:
        ws_server = get_server()
        if not ws_server:
            return {"success": False, "message": "WebSocket server not initialized"}
        
        try:
            # Execute the original function to get action name and data
            action_name, action_data = await func(*args, **kwargs)
            
            # Send to plugin
            try:
                response = await ws_server.send_instruction(
                    action=action_name, 
                    data=action_data
                )
            except TimeoutError:
                return {"success": False, "message": "Timeout waiting for response from plugin"}
            
            # Process response
            if not response:
                return {"success": False, "message": "No response from plugin"}
            
            logger.info(f"Received response: {response}")
            if "success" in response and response["success"]:
                return {
                    "success": True,
                    "response": response.get("result", {})
                }
            else:
                return {"success": False, "message": response.get("error", "Unknown error")}
                
        except Exception as e:
            logger.exception(f"Error in {func.__name__}:")
            return {"success": False, "message": f"Error in {func.__name__}: {str(e)}"}
    
    return wrapper

# Project management action definitions
class PremiereProjectActions:
    """Implementation of Premiere Pro project-related actions"""
    
    @staticmethod
    @premiere_action
    async def get_active(**_) -> Tuple[str, Dict[str, Any]]:
        """Get active project (extra arguments are ignored)"""
        return "getActiveProject", {}
    
    @staticmethod
    @premiere_action
    async def open_project(path: str, **options) -> Tuple[str, Dict[str, Any]]:
        """Open a project"""
        return "openProject", {"path": path, "options": options}
    
    @staticmethod
    @premiere_action
    async def create_project(path: str, **_) -> Tuple[str, Dict[str, Any]]:
        """Create a new project"""
        return "createProject", {"path": path}
    
    @staticmethod
    @premiere_action
    async def save_project(**_) -> Tuple[str, Dict[str, Any]]:
        """Save the project"""
        return "saveProject", {}
    
    @staticmethod
    @premiere_action
    async def save_project_as(path: str, **_) -> Tuple[str, Dict[str, Any]]:
        """Save the project with a new name"""
        return "saveProjectAs", {"path": path}
    
    @staticmethod
    @premiere_action
    async def close_project(**options) -> Tuple[str, Dict[str, Any]]:
        """Close the project"""
        return "closeProject", {"options": options}

# Sequence management action definitions
class PremiereSequenceActions:
    """Implementation of Premiere Pro sequence-related actions"""
    
    @staticmethod
    @premiere_action
    async def get_active(**_) -> Tuple[str, Dict[str, Any]]:
        """Get active sequence"""
        return "getActiveSequence", {}
    
    @staticmethod
    @premiere_action
    async def create_sequence(name: str, preset_path: Optional[str] = None, **_) -> Tuple[str, Dict[str, Any]]:
        """Create a new sequence"""
        return "createSequence", {"name": name, "presetPath": preset_path}
    
    @staticmethod
    @premiere_action
    async def create_sequence_from_media(
        name: str, 
        clip_project_items: list, 
        target_bin: Optional[str] = None,
        **_
    ) -> Tuple[str, Dict[str, Any]]:
        """Create a new sequence from media"""
        return "createSequenceFromMedia", {
            "name": name, 
            "clipProjectItems": clip_project_items,
            "targetBin": target_bin
        }
    
    @staticmethod
    @premiere_action
    async def set_active_sequence(sequence_id: str, **_) -> Tuple[str, Dict[str, Any]]:
        """Set active sequence"""
        return "setActiveSequence", {"sequenceId": sequence_id}
    
    @staticmethod
    @premiere_action
    async def get_sequence_list(**_) -> Tuple[str, Dict[str, Any]]:
        """Get list of all sequences in the project"""
        return "getSequenceList", {}
    
    @staticmethod
    @premiere_action
    async def get_player_position(**_) -> Tuple[str, Dict[str, Any]]:
        """Get playhead position"""
        return "getPlayerPosition", {}
    
    @staticmethod
    @premiere_action
    async def set_player_position(position, **_) -> Tuple[str, Dict[str, Any]]:
        """Set playhead position"""
        return "setPlayerPosition", {"position": position}

# Media import action definitions
class PremiereMediaActions:
    """Implementation of Premiere Pro media-related actions"""
    
    @staticmethod
    @premiere_action
    async def import_files(
        file_paths: list, 
        suppress_ui: bool = False, 
        target_bin: Optional[str] = None,
        as_numbered_stills: bool = False,
        **_
    ) -> Tuple[str, Dict[str, Any]]:
        """Import files"""
        return "importFiles", {
            "filePaths": file_paths,
            "suppressUI": suppress_ui,
            "targetBin": target_bin,
            "asNumberedStills": as_numbered_stills
        }
    
    @staticmethod
    @premiere_action
    async def import_sequences(
        project_path: str, 
        sequence_ids: list,
        **_
    ) -> Tuple[str, Dict[str, Any]]:
        """Import sequences from another project"""
        return "importSequences", {
            "projectPath": project_path,
            "sequenceIds": sequence_ids
        }
    
    @staticmethod
    @premiere_action
    async def import_ae_comps(
        aep_path: str, 
        comp_names: list, 
        target_bin: Optional[str] = None,
        **_
    ) -> Tuple[str, Dict[str, Any]]:
        """Import After Effects compositions"""
        return "importAEComps", {
            "aepPath": aep_path,
            "compNames": comp_names,
            "targetBin": target_bin
        }
    
    @staticmethod
    @premiere_action
    async def import_all_ae_comps(
        aep_path: str, 
        target_bin: Optional[str] = None,
        **_
    ) -> Tuple[str, Dict[str, Any]]:
        """Import all After Effects compositions"""
        return "importAllAEComps", {
            "aepPath": aep_path,
            "targetBin": target_bin
        }

# Registration with MCP Server
def register_premiere_tools(server: Any) -> None:
    """
    Register Premiere Pro tools with the MCP server
    
    Args:
        server: MCP server instance
    """
    logger.info("Registering Premiere Pro tools")
    
    # Project management tool
    @server.tool()
    async def manage_project(action: str = "get_active", **kwargs) -> Dict[str, Any]:
        """
        Manage Adobe Premiere Pro projects
        
        Args:
            action: Action to execute
                - get_active: Get information about the active project
                - open_project: Open a project (path, options)
                - create_project: Create a new project (path)
                - save_project: Save the project
                - save_project_as: Save the project with a new name (path)
                - close_project: Close the project (options)
            **kwargs: Additional parameters according to the action
            
        Returns:
            Dict[str, Any]: Operation result
        """
        # Convert action name to method name (get_active → get_active)
        method_name = action.lower()
        
        # Action dispatch
        action_method = getattr(PremiereProjectActions, method_name, None)
        if not action_method:
            return {"success": False, "message": f"Unknown project action: {action}"}
            
        return await action_method(**kwargs["kwargs"])
    
    # Sequence management tool
    @server.tool()
    async def manage_sequence(action: str = "get_active", **kwargs) -> Dict[str, Any]:
        """
        Manage Adobe Premiere Pro sequences
        
        Args:
            action: Action to execute
                - get_active: Get information about the active sequence
                - create_sequence: Create a new sequence (name, preset_path)
                - create_sequence_from_media: Create a new sequence from media (name, clip_project_items, target_bin)
                - set_active_sequence: Set active sequence (sequence_id)
                - get_sequence_list: Get list of all sequences in the project
                - get_player_position: Get playhead position
                - set_player_position: Set playhead position (position)
            **kwargs: Additional parameters according to the action
            
        Returns:
            Dict[str, Any]: Operation result
        """
        # Convert action name to method name (get_active → get_active)
        method_name = action.lower()
        
        # Action dispatch
        action_method = getattr(PremiereSequenceActions, method_name, None)
        if not action_method:
            return {"success": False, "message": f"Unknown sequence action: {action}"}
            
        logger.info(f"Executing sequence action: {action} with params: {kwargs['kwargs']}")
        return await action_method(**kwargs["kwargs"])
    
    # Media management tool
    @server.tool()
    async def manage_media(action: str, **kwargs) -> Dict[str, Any]:
        """
        Manage Adobe Premiere Pro media
        
        Args:
            action: Action to execute
                - import_files: Import files (file_paths, suppress_ui, target_bin, as_numbered_stills)
                - import_sequences: Import sequences from another project (project_path, sequence_ids)
                - import_ae_comps: Import AE compositions (aep_path, comp_names, target_bin)
                - import_all_ae_comps: Import all AE compositions (aep_path, target_bin)
            **kwargs: Additional parameters according to the action
            
        Returns:
            Dict[str, Any]: Operation result
        """
        # Convert action name to method name (import_files → import_files)
        method_name = action.lower()
        
        # Action dispatch
        action_method = getattr(PremiereMediaActions, method_name, None)
        if not action_method:
            return {"success": False, "message": f"Unknown media action: {action}"}
            
        return await action_method(**kwargs["kwargs"])

File: plugin/src/tools.ts
// import type { premierepro, Project, Sequence } from "../types.d.ts";
const app = require("premierepro") // as premierepro;
/**
 * Adobe Premiere Pro API Extension Module
 * Provides a version with consistent response format
 */

// Standard response format
interface StandardResponse {
  success: boolean;
  data?: any;
  error?: string;
}

/**
 * Execute the specified action
 */
export async function executeAction(action: string, data?: any): Promise<StandardResponse> {
    try {
        switch (action) {
            // Project related
            case 'getActiveProject':
                return await getActiveProject();
            case 'openProject':
                return await openProject(data.path, data.options);
            case 'createProject':
                return await createProject(data.path);
            case 'saveProject':
                return await saveProject();
            case 'saveProjectAs':
                return await saveProjectAs(data.path);
            case 'closeProject':
                return await closeProject(data.options);
                
            // Sequence related
            case 'getActiveSequence':
                return await getActiveSequence();
            case 'createSequence':
                return await createSequence(data.name, data.presetPath);
            case 'createSequenceFromMedia':
                return await createSequenceFromMedia(data.name, data.clipProjectItems, data.targetBin);
            case 'setActiveSequence':
                return await setActiveSequence(data.sequenceId);
            case 'getSequenceList':
                return await getSequenceList();
            case 'getPlayerPosition':
                return await getPlayerPosition();
            case 'setPlayerPosition':
                return await setPlayerPosition(data.position);
                
            // Media import related
            case 'importFiles':
                return await importFiles(data.filePaths, data.suppressUI, data.targetBin, data.asNumberedStills);
            case 'importSequences':
                return await importSequences(data.projectPath, data.sequenceIds);
            case 'importAEComps':
                return await importAEComps(data.aepPath, data.compNames, data.targetBin);
            case 'importAllAEComps':
                return await importAllAEComps(data.aepPath, data.targetBin);
                
            default:
                return {
                    success: false,
                    error: `Unknown action: ${action}`
                };
        }
    } catch (error) {
        return {
            success: false,
            error: error instanceof Error ? error.message : String(error)
        };
    }
}

/**
 * Wrapper function for error handling
 */
async function safeExecute<T>(callback: () => Promise<T>): Promise<StandardResponse> {
    try {
        const result = await callback();
        return {
            success: true,
            data: result
        };
    } catch (error) {
        console.error("An error occurred:", error);
        console.error("Error stack trace:\n", error.stack);
        return {
            success: false,
            error: error instanceof Error ? error.message : String(error)
        };
    }
}

/**
 * === Project Related Functions ===
 */

/**
 * Get information about the active project
 */
async function getActiveProject(): Promise<StandardResponse> {
    return safeExecute(async () => {
        const project = await app.Project.getActiveProject();
        if (!project) {
            throw new Error('No active project');
        }
        return {
            name: project.name,
            path: project.path,
            id: project.guid.toString()
        };
    });
}

/**
 * Open a project
 */
async function openProject(path: string, options?: any): Promise<StandardResponse> {
    return safeExecute(async () => {
        let openOptions = null;
        
        if (options) {
            openOptions = {};
            if (options.showConvertProjectDialog !== undefined) {
                openOptions.setShowConvertProjectDialog(options.showConvertProjectDialog);
            }
            if (options.showLocateFileDialog !== undefined) {
                openOptions.setShowLocateFileDialog(options.showLocateFileDialog);
            }
            if (options.showWarningDialog !== undefined) {
                openOptions.setShowWarningDialog(options.showWarningDialog);
            }
            if (options.addToMRUList !== undefined) {
                openOptions.setAddToMRUList(options.addToMRUList);
            }
        }
        
        const project = await app.Project.open(path, openOptions);
        return {
            name: project.name,
            path: project.path,
            id: project.guid.toString()
        };
    });
}

/**
 * Create a new project
 */
async function createProject(path: string): Promise<StandardResponse> {
    return safeExecute(async () => {
        const project = await app.Project.createProject(path);
        return {
            name: project.name,
            path: project.path,
            id: project.guid.toString()
        };
    });
}

/**
 * Save the active project
 */
async function saveProject(): Promise<StandardResponse> {
    return safeExecute(async () => {
        const project = await app.Project.getActiveProject();
        if (!project) {
            throw new Error('No active project');
        }
        
        const result = await project.save();
        // Always return as an object instead of directly returning the boolean value
        return {
            saved: result
        };
    });
}

/**
 * Save the active project to a specified path
 */
async function saveProjectAs(path: string): Promise<StandardResponse> {
    return safeExecute(async () => {
        const project = await app.Project.getActiveProject();
        if (!project) {
            throw new Error('No active project');
        }
        
        const result = await project.saveAs(path);
        return {
            saved: result,
            path: path
        };
    });
}

/**
 * Close the active project
 */
async function closeProject(options?: any): Promise<StandardResponse> {
    return safeExecute(async () => {
        const project = await app.Project.getActiveProject();
        if (!project) {
            throw new Error('No active project');
        }
        
        let closeOptions = null;
        
        if (options) {
            closeOptions = {};
            if (options.promptIfDirty !== undefined) {
                closeOptions.setPromptIfDirty(options.promptIfDirty);
            }
            if (options.showCancelButton !== undefined) {
                closeOptions.setShowCancelButton(options.showCancelButton);
            }
            if (options.isAppBeingPreparedToQuit !== undefined) {
                closeOptions.setIsAppBeingPreparedToQuit(options.isAppBeingPreparedToQuit);
            }
            if (options.saveWorkspace !== undefined) {
                closeOptions.setSaveWorkspace(options.saveWorkspace);
            }
        }
        
        const result = await project.close(closeOptions);
        return {
            closed: result
        };
    });
}

/**
 * === Sequence Related Functions ===
 */

/**
 * Get information about the active sequence
 */
async function getActiveSequence(): Promise<StandardResponse> {
    return safeExecute(async () => {
        const project = await app.Project.getActiveProject();
        if (!project) {
            throw new Error('No active project');
        }
        
        const sequence = await project.getActiveSequence();
        if (!sequence) {
            throw new Error('No active sequence');
        }
        
        return {
            name: sequence.name,
            id: sequence.guid.toString(),
            videoTrackCount: await sequence.getVideoTrackCount(),
            audioTrackCount: await sequence.getAudioTrackCount(),
            captionTrackCount: await sequence.getCaptionTrackCount()
        };
    });
}

/**
 * Create a new sequence
 */
async function createSequence(name: string, presetPath?: string): Promise<StandardResponse> {
    return safeExecute(async () => {
        const project = await app.Project.getActiveProject();
        if (!project) {
            throw new Error('No active project');
        }
        
        const sequence = await project.createSequence(name, presetPath);
        return {
            name: sequence.name,
            id: sequence.guid.toString()
        };
    });
}

/**
 * Create a sequence from selected media
 */
async function createSequenceFromMedia(name: string, clipProjectItems: any[], targetBin?: any): Promise<StandardResponse> {
    return safeExecute(async () => {
        const project = await app.Project.getActiveProject();
        if (!project) {
            throw new Error('No active project');
        }
        
        const sequence = await project.createSequenceFromMedia(name, clipProjectItems, targetBin);
        return {
            name: sequence.name,
            id: sequence.guid.toString()
        };
    });
}

/**
 * Set a sequence as active
 */
async function setActiveSequence(sequenceId: string): Promise<StandardResponse> {
    return safeExecute(async () => {
        const project = await app.Project.getActiveProject();
        if (!project) {
            throw new Error('No active project');
        }
        
        // Create Guid object
        console.debug("[DEBUG] sequenceId:", sequenceId);
        const guid = app.Guid.fromString(sequenceId);
        if (!guid) {
            throw new Error(`Invalid sequence ID: ${sequenceId}`);
        }
        
        console.debug("[DEBUG] guid:", guid);
        const sequence = await project.getSequence(guid);
        if (!sequence) {
            throw new Error(`Sequence with ID ${sequenceId} not found`);
        }
        
        console.debug("[DEBUG] sequence:", sequence);
        // Set sequence as active
        const result = await project.setActiveSequence(sequence);
        console.debug("[DEBUG] result:", result);
        return {
            activated: result,
            sequenceId: sequenceId
        };
    });
}

/**
 * Get a list of all sequences in the project
 */
async function getSequenceList(): Promise<StandardResponse> {
    return safeExecute(async () => {
        const project = await app.Project.getActiveProject();
        if (!project) {
            throw new Error('No active project');
        }
        
        const sequences = await project.getSequences();
        const sequenceList = sequences.map((seq: any) => ({
            name: seq.name,
            id: seq.guid.toString()
        }));
        
        // Always wrap the list in a dictionary instead of returning it directly
        return {
            sequences: sequenceList
        };
    });
}

/**
 * Get the current playhead position
 */
async function getPlayerPosition(): Promise<StandardResponse> {
    return safeExecute(async () => {
        const project = await app.Project.getActiveProject();
        if (!project) {
            throw new Error('No active project');
        }
        
        const sequence = await project.getActiveSequence();
        if (!sequence) {
            throw new Error('No active sequence');
        }
        
        const position = await sequence.getPlayerPosition();
        return {
            seconds: position.seconds,
            ticks: position.ticks
        };
    });
}

/**
 * Set the playhead position
 */
async function setPlayerPosition(position: any): Promise<StandardResponse> {
    return safeExecute(async () => {
        const project = await app.Project.getActiveProject();
        if (!project) {
            throw new Error('No active project');
        }
        
        const sequence = await project.getActiveSequence();
        if (!sequence) {
            throw new Error('No active sequence');
        }
        
        let tickTime;
        if (typeof position === 'number') {
            // If specified in seconds
            tickTime = app.TickTime.createWithSeconds(position);
        } else if (position.ticks) {
            // If specified in ticks
            tickTime = app.TickTime.createWithTicks(position.ticks);
        } else if (position.frame !== undefined && position.frameRate !== undefined) {
            // If specified in frames and frame rate
            const frameRate = app.FrameRate.createWithValue(position.frameRate);
            tickTime = app.TickTime.createWithFrameAndFrameRate(position.frame, frameRate);
        } else {
            throw new Error('Invalid position format');
        }
        
        const result = await sequence.setPlayerPosition(tickTime);
        return {
            positioned: result,
            seconds: tickTime.seconds
        };
    });
}

/**
 * === Media Import Related Functions ===
 */

/**
 * Import files into the project
 */
async function importFiles(filePaths: string[], suppressUI?: boolean, targetBin?: any, asNumberedStills?: boolean): Promise<StandardResponse> {
    return safeExecute(async () => {
        const project = await app.Project.getActiveProject();
        if (!project) {
            throw new Error('No active project');
        }
        
        const result = await project.importFiles(filePaths, suppressUI, targetBin, asNumberedStills);
        return {
            imported: result,
            fileCount: filePaths.length,
            files: filePaths
        };
    });
}

/**
 * Import sequences from another project
 */
async function importSequences(projectPath: string, sequenceIds: string[]): Promise<StandardResponse> {
    return safeExecute(async () => {
        const project = await app.Project.getActiveProject();
        if (!project) {
            throw new Error('No active project');
        }
        
        const result = await project.importSequences(projectPath, sequenceIds);
        return {
            imported: result,
            sequenceCount: sequenceIds.length,
            sourceProject: projectPath
        };
    });
}

/**
 * Import After Effects compositions
 */
async function importAEComps(aepPath: string, compNames: string[], targetBin?: any): Promise<StandardResponse> {
    return safeExecute(async () => {
        const project = await app.Project.getActiveProject();
        if (!project) {
            throw new Error('No active project');
        }
        
        const result = await project.importAEComps(aepPath, compNames, targetBin);
        return {
            imported: result,
            compCount: compNames.length,
            aepPath: aepPath
        };
    });
}

/**
 * Import all After Effects compositions from a project
 */
async function importAllAEComps(aepPath: string, targetBin?: any): Promise<StandardResponse> {
    return safeExecute(async () => {
        const project = await app.Project.getActiveProject();
        if (!project) {
            throw new Error('No active project');
        }
        
        const result = await project.importAllAEComps(aepPath, targetBin);
        return {
            imported: result,
            aepPath: aepPath
        };
    });
}
